### PR TITLE
[lld-macho][NFC] Refactor insertions into inputSections

### PR DESCRIPTION
**Pull request by alx32**
Before this change, after `InputSection` objects are created, they need to be added to the appropriate container for tracking.
The logic for selecting the appropriate container lives in `Driver.cpp` / `gatherInputSections`, where the `InputSection` is added to the matching container depending on the input config and the  type of  `InputSection`.

Also, multiple other locations also insert directly into `inputSections` array - assuming that that is the appropriate container for the `InputSection`'s they create. Currently this is the correct assumption, however an upcoming feature will change this. 

For an upcoming feature (relative method lists), we need to route `InputSection`'s either to `inputSections` array or to a synthetic section, depending on weather the relative method list optimization is enabled or not.

We can achieve the above either by duplicating some of the logic or refactoring the routing and  `InputSection`'s and reusing that. 

The refactoring & code sharing approach seems the correct way to go - as such this diff performs the refactoring while not introducing any functional changes. Later on we can just call `addInputSection` and not have to worry about routing logic.

### CODE REVIEW

File: lld/MachO/Driver.cpp [hunk 1]
```diff
@@ -1220,53 +1220,18 @@ static void createFiles(const InputArgList &args) {
...
 static void gatherInputSections() {
   TimeTraceScope timeScope("Gathering input sections");
-  int inputOrder = 0;
   for (const InputFile *file : inputFiles) {
     for (const Section *section : file->sections) {
       // Compact unwind entries require special handling elsewhere. (In
       // contrast, EH frames are handled like regular ConcatInputSections.)
       if (section->name == section_names::compactUnwind)
         continue;
-      ConcatOutputSection *osec = nullptr;
```
Comment (alx32) [inline lld/MachO/Driver.cpp:1230]: Diff vs code added below: 

image

File: lld/MachO/InputSection.cpp [hunk 2]
```diff
@@ -38,6 +38,43 @@ static_assert(sizeof(void *) != 8 ||
...
+  // Used across function calls to impose section ordering
+  static uint64_t inputOrder = 0;
+
+  if (auto *isec = dyn_cast<ConcatInputSection>(inputSection)) {
+    if (isec->isCoalescedWeak())
+      return;
+    if (config->emitInitOffsets &&
+        sectionType(isec->getFlags()) == S_MOD_INIT_FUNC_POINTERS) {
+      in.initOffsets->addInput(isec);
+      return;
```
Comment (ellishg) [inline lld/MachO/InputSection.cpp:51]: These returns will skip the `assert()` below. Is that ok?

Comment (alx32) [reply lld/MachO/InputSection.cpp:51]: Yes, we only modify inputOrder past the return's.

File: lld/MachO/InputSection.cpp [hunk 3]
```diff
@@ -38,6 +38,43 @@ static_assert(sizeof(void *) != 8 ||
 
 std::vector<ConcatInputSection *> macho::inputSections;
 
+void lld::macho::addInputSection(InputSection *inputSection) {
+  // Used across function calls to impose section ordering
+  static uint64_t inputOrder = 0;
```
Comment (alx32) [inline lld/MachO/InputSection.cpp:43]: TODO: make inputOrder `int`, as original, not `uint64_t`

[the remark this answers is not in the document]
Comment (alx32) [reply lld/MachO/InputSection.cpp:43]: How about defining it next to `inputSections` and resetting it in `cleanupCallback` along with `inputSections` ?

File: lld/MachO/InputSection.h [hunk 4]
```diff
@@ -369,6 +369,7 @@ constexpr const char addrSig[] = "__llvm_addrsig";
 
 } // namespace section_names
 
+void addInputSection(InputSection *inputSection);
```
[the remark this answers is not in the document]
Comment (alx32) [reply lld/MachO/InputSection.h:374]: Added docs to function body.

**Review by kyulee-com — APPROVED** [review]
lgtm

**Pull request merged**